pkg/exchange/coinbase/exchage.go

```diff
@@ -0,0 +1,309 @@
+	cbOrders, err := e.queryOrdersByPagination(ctx, toLocalSymbol(symbol), []string{"open"})
+	if err != nil {
+		return nil, err
+	}
+	orders := make([]types.Order, 0, len(cbOrders))
+	for _, cbOrder := range cbOrders {
+		orders = append(orders, *toGlobalOrder(&cbOrder))
+	}
+	return orders, nil
+}
+
+func (e *Exchange) queryOrdersByPagination(ctx context.Context, symbol string, status []string) ([]api.Order, error) {
+	sortedBy := "created_at"
+	sorting := "desc"
+	before := time.Now()
+	localSymbol := toLocalSymbol(symbol)
+	paginationLimit := 1000
+	getOrdersReq := e.client.NewGetOrdersRequest()
+	getOrdersReq.ProductID(localSymbol).Status(status).SortedBy(sortedBy).Sorting(sorting).Before(before).Limit(paginationLimit)
+	cbOrders, err := getOrdersReq.Do(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get orders")
+	}
+
+	if len(cbOrders) < paginationLimit {
+		return cbOrders, nil
+	}
+
+	done := false
+	for {
+		if done {
+			break
+		}
+
+		before = time.Time(cbOrders[len(cbOrders)-1].CreatedAt)
+		getOrdersReq.Before(before)
+		newOrders, err := getOrdersReq.Do(ctx)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to get orders while paginating")
+		}
+		if len(newOrders) < paginationLimit {
+			done = true
+		}
+		cbOrders = append(cbOrders, newOrders...)
+	}
+	return cbOrders, nil
+}
+
+func (e *Exchange) CancelOrders(ctx context.Context, orders ...types.Order) error {
+	failedOrderIDs := make([]string, 0)
+	for _, order := range orders {
+		req := e.client.NewCancelOrderRequest().OrderID(order.UUID)
+		res, err := req.Do(ctx)
+		if err != nil {
+			log.WithError(err).Errorf("failed to cancel order: %v", order.UUID)
+			failedOrderIDs = append(failedOrderIDs, order.UUID)
+		}
+		log.Infof("order %v has been cancelled", res)
+	}
+	if len(failedOrderIDs) > 0 {
+		return errors.Errorf("failed to cancel orders: %v", failedOrderIDs)
+	}
+	return nil
+}
+
+// ExchangeMarketDataService
+func (e *Exchange) NewStream() types.Stream {
+	return nil
+}
+
+func (e *Exchange) QueryMarkets(ctx context.Context) (types.MarketMap, error) {
+	req := e.client.NewGetMarketInfoRequest()
+	markets, err := req.Do(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get markets")
+	}
+	marketMap := make(types.MarketMap)
+	for _, m := range markets {
+		marketMap[toGlobalSymbol(m.ID)] = *toGlobalMarket(&m)
+	}
+	return marketMap, nil
+}
+
+func (e *Exchange) QueryTicker(ctx context.Context, symbol string) (*types.Ticker, error) {
+	req := e.client.NewGetTickerRequest().ProductID(toLocalSymbol(symbol))
+	ticker, err := req.Do(ctx)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get ticker: %v", symbol)
+	}
+	return toGlobalTicker(ticker), nil
+}
+
+func (e *Exchange) QueryTickers(ctx context.Context, symbol ...string) (map[string]types.Ticker, error) {
+	tickers := make(map[string]types.Ticker)
+	for _, s := range symbol {
+		ticker, err := e.QueryTicker(ctx, s)
```

Comment:
does coinbase support multiple ticker query in one request?

Reply:
not likely.

I do not see it in the api reference:
https://docs.cdp.coinbase.com/exchange/reference/exchangerestapi_getproductticker

It states that if you want ticks of multiple products, it's recommended to do it by connecting to websocket feeds.